### ISSUE TITLE
site_prisms has runtime dependency - capybara ~> 2.12, is support for capybara 3.0.2 expected

### ISSUE THREAD

**Issue by crcraig92** [natritmeyer/site_prism#242]
# Issue to raise for SitePrism framework
I'm not sure I have all of the information you would like for this.  I'm mainly just inquiring as to why site_prism has the gemspec locking Capybara versions ~> 2.12

## Environment
- Windows 10
- Site_Prism 2.12 and Capybara 2.18.0 (hoping to use 3.0.2) (Also selenium 3.11.0)
- Ruby 2.3.3p222
- Cucumber tests mainly (cucumber-core 3.1.0)
- Chrome 66 (chromedriver 2.37), Firefox 59.0.2 (geckodriver 0.20.1)

## Expected Behavior
Expected when I update capybara to 3.0.2, site_prism either updates or stays at the current version

## Actual Behavior
site_prism downgrades to version 1.3 and I begin getting errors with the wrong number of arguments

I have not looked into the argument errors, to be honest.  I started to and realized that the version had dropped significantly and I don't want to go back in time to before new features and support was added.

## Steps to reproduce
Bundle Update?

I'm not sure how relevant all of these fields are for an issue of this nature.  I'll do my best to assist on this issue.

**Comment by tgaff** [comment]
Thanks for filing this @crcraig92 

I just ran a quick pass through of the tests with capybara 3.0.2.  I didn't encounter anything unexpected so this may end up being pretty quick.

**Comment by tgaff** [comment]
I have a partially completed change for this.  I ran into a bug while working on it though.  That resulted in PR #243 which will need to be dealt with first.

**Comment by luke-hill** [comment]
@crcraig92 To answer your question why site_prism is locked to 2.12. Due to the way semantic versioning works (See semver.org), we cannot guarantee support for the 3.x series of capybara.

However, what we can do is look into it (As @tgaff) has done above.

Main concerns / points to note.

Ruby Version / XPath version bump (Extraneous)
first / all methods have been reworked, will need some additional work.
Generic testing around refactors done (<100 commits so not too worried)
Bumps through the 2.x series (2.12 - 2.18 includes several hundred commits)

If we did land this, I would probably want to halve the enforced capybara updates first (To something around 2.15), to ensure people are being staged with their updates.

**Comment by crcraig92** [comment]
@tgaff Thanks for taking an initial look into this effort

@luke-hill Thanks for the information!  That makes sense.  I appreciate your efforts to maintain this gem.

**Comment by luke-hill** [comment]
On a side note. At the moment you're not too far behind Capybara using 2.18.0 So hopefully being on 2.18.0 isn't that much of a hindrance.

We definitely will get 3.x unlocked, I promise. Just a question of when, not if.

**Comment by luke-hill** [comment]
I've popped on a "Needs decision" label @crcraig92 purely for the reason that "we" or "I" need to decide "how" this will land.

My personal opinion is.
a) It cannot land as the next release
b) It should be staged (Even through fake upped dependencies)
c) (Unsure if we can), We should see what the uptake is RE Ruby version / Capybara version (I know Capybara isn't high on the 3.x usage at the moment, although that might just be because of individual gemfile restrictions. From this we can then decide when to support this.
d) If we do decide to include this, we should probably find out from Capybara maintainers, if there are any other decent changes we can make to facilitate it.
e) If this lands, would we want it to be a breaking change version (And correlate a V3.0 site_prism release accordingly), or are we happy having this in site_prism 2.x

**Comment by luke-hill** [comment]
Ping @tmertens @twapole @TheMetalCode for additional thoughts / 2cents.

**Comment by twalpole** [comment]
@luke-hill I haven't looked through any recent updates on this project, so this may have already been handled, but I'm pretty sure the behavioral changes in Capybaras `first` and `all` would require at least some changes to SitePrism to maintain its current behaviors around waiting/not-waiting

**Comment by luke-hill** [comment]
Cheers. In my eyes I want this project to be as conformant to Capybara as possible going forward. It makes sense given the mass usage that this gem should just wrap your behaviour and not change that much to be honest. (Aside from adding stuff).

**Comment by crcraig92** [comment]
I'm not sure if you were asking me for my input on this.  But here is my opinions in regards to yours anyway.

> a) It cannot land as the next release

It's not really breaking anything, so there's no need to rush it.  Cucumber moved to 3.x recently and made some big changes and shortly after Capybara also went to 3.x and push some changes I've been looking forward to.  .find now returns ElementNotFound instead of nil and .all has some implicit waits now.

> b) It should be staged (Even through fake upped dependencies)

This sounds good to me.

> c) (Unsure if we can), We should see what the uptake is RE Ruby version / Capybara version (I know Capybara isn't high on the 3.x usage at the moment, although that might just be because of individual gemfile restrictions. From this we can then decide when to support this.

I'm using Ruby 2.3.3, and I've tested 2.4 for my framework and it's fine.  Some other frameworks are stuck at 2.3.3 for 'reasons' and ruby version management isn't the cleanest thing on Windows.  I would have thought a larger portion of Capybara users would also be using Site Prism but it appears that's not the case.  I'm not sure what other gems might have dependency issues, but for me I believe the only gem restricting the bundler from grabbing Capybara 3.x is Site Prism.  

> d) If we do decide to include this, we should probably find out from Capybara maintainers, if there are any other decent changes we can make to facilitate it.

As @twalpole mentioned, first and all had significant changes, also Capybara.exact_options might affect users depending on how they created their finders.  I can't think of anything else that would need to be handled.

> e) If this lands, would we want it to be a breaking change version (And correlate a V3.0 site_prism release accordingly), or are we happy having this in site_prism 2.x

I'd say it should move to 3.x if at the very least just for clarity.  Selenium, Cucumber, and Capybara are in a 3.x series so it seems like it would fit.  Also, if you need to make changes to implicit/explicit waits people need to pay more attention to updating or they may start seeing timing issues.

I hope that answers any questions if you had them, and didn't generate more questions/concerns you didn't have before.  

Thanks again for your efforts with this.

**Comment by tgaff** [comment]
I've been looking at the first/all changes.  I don't think they'll be particularly difficult to take care of.  Those are the main ones mentioned in https://github.com/teamcapybara/capybara/blob/master/UPGRADING.md    Capybara 3.0 doesn't look like a huge departure from 2.x.  
The only reason I haven't prepped a PR for this already is because of other blocking issues that should be resolved and merged first.

I also don't think there's a big need to lock site_prism to a last capybara 2.x and then have a new 3.x supporting version.  From what I've looked at so far it seems likely that we can support 2.x and 3.x simultaneously without significant effort.  I think it should be up to the user - if they have specific requirements in a page object that don't work in Capybara 2.x then the user will be responsible for choosing when to switch to Capybara 3.x and make the required changes.

**Comment by luke-hill** [comment]
This will form the basis of a 2.14 site_prism release.

2.13 will be released shortly (Around 1-2 weeks). This will then (Whenever complete), hit the following cut.

**Comment by danielricecodes** [comment]
For anyone still waiting for the officially cut gem, you can use a fork with slightly modified dependencies:

```ruby
gem 'site_prism', github: 'mateusg/site_prism', ref: 'dc00f2b976e0f801541589ff94ded23cb04cee02'
```

**Comment by luke-hill** [comment]
Fixed in fb182c8eec3d4b689bd906200f9432db6368b685 - Will be released in the next 24 hours as v2.15 (Note there are just fixes in here aside from the uncapping dependencies, just to inform readers that it wasn't as simple as just upping the gemspec requirements)
